feat(zone.js): add awaiter transform spec implementation

JiaLiPassion opened the pull request:
The async+await syntax is native and there is no hook for Zone.js to
patch. So it cannot intercept the callbacks when awaited functions
complete.

Angular CLI now only transform `async/await` to `Javascript Promise`,
but it is still not idle, this PR did more enhancement to generate
native `async/await` and `generator` code.

The benefits are:

1. The performance is better.
2. Debug experience is better.

To use this feature. There are 2 use cases:

1. With Angular CLI, in this PR, `zone.js` provides a new entry point
`zone.js/awaiter`, this bundle is a `es2017` bundle. Angular CLI will
import this bundle and also transform the `async/await` code to
`generator` like this.

```
async function testAsync() {
  let a = 0;
  a++;
  console.log('before await', a, Zone.current.name);
  a = await compute(a);
  a++;
  console.log('after await', a, Zone.current.name);
}
```

```
function testAsync_generator_1* () {
  let a = 0;
  a++;
  console.log('before await', a, Zone.current.name);
  a = yield compute(a);
  a++;
  console.log('after await', a, Zone.current.name);
}
```

2. Without Angular CLI, the user need to import `zone.js/awaiter` after
import `zone.js` themselves, and also need to do the transform with
their own build pipeline.

JiaLiPassion committed:
feat(zone.js): add awaiter transform spec implementation

The async+await syntax is native and there is no hook for Zone.js to
patch. So it cannot intercept the callbacks when awaited functions
complete.

Angular CLI now only transform `async/await` to `Javascript Promise`,
but it is still not idle, this PR did more enhancement to generate
native `async/await` and `generator` code.

The benefits are:

1. The performance is better.
2. Debug experience is better.

To use this feature. There are 2 use cases:

1. With Angular CLI, in this PR, `zone.js` provides a new entry point
`zone.js/awaiter`, this bundle is a `es2017` bundle. Angular CLI will
import this bundle and also transform the `async/await` code to
`generator` like this.

```
async function testAsync() {
  let a = 0;
  a++;
  console.log('before await', a, Zone.current.name);
  a = await compute(a);
  a++;
  console.log('after await', a, Zone.current.name);
}
```

```
function testAsync_generator_1* () {
  let a = 0;
  a++;
  console.log('before await', a, Zone.current.name);
  a = yield compute(a);
  a++;
  console.log('after await', a, Zone.current.name);
}
```

2. Without Angular CLI, the user need to import `zone.js/awaiter` after
import `zone.js` themselves, and also need to do the transform with
their own build pipeline.

petebacondarwin commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -861,6 +904,14 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+  };
 }, true);
+
+Zone.__load_patch('fake_async_awaiter', (global, Zone, api) => {});
```
Also, it looks like you ran "Prettier" on this file, which is cool. But it makes the changes for this particular feature difficult to spot. I wonder if you could move the formatting changes to a separate commit?

JiaLiPassion replied:
This line is not needed, I just deleted it, also I use `clang-format` to re-format this file.

petebacondarwin commented on packages/zone.js/test/native/native-async-await.common.spec.ts:
```diff
@@ -0,0 +1,505 @@
+                     }));
+}
+
+async function testAsync1(options: {a: number; throwError: boolean} = {
+  a: 0,
+  throwError: false
+}) {
+  let a = options.a;
+  a++;
+  a = await compute(a);
+  a++;
+  if (options.throwError) {
+    throw new Error('throw error in async function');
+  }
+  a = await compute(a);
+  a++;
+  return a;
+}
```
So as far as I can see, this function is only here as a reminder of what the generated code (in `generatorTest1()`) is supposed to represent, yes? Is it actually needed?

petebacondarwin replied:
Similarly for `testAsync2`.

Perhaps just move them to comment blocks?

JiaLiPassion replied:
Yes, you are right, I moved the code to comments.

petebacondarwin commented on packages/zone.js/test/native/native-async-await.common.spec.ts:
```diff
@@ -0,0 +1,505 @@
+        });
+      });
+    });
+    it('should run into the original zone after await without parameter', (done: DoneFn) => {
```
Is this `it()` description accurate?
This spec seems to be testing a nested generator scenario.

petebacondarwin replied:
Oh, I see. I think it is testing the same scenario as the spec on line 134 above, but driving it via an outer `__awaiter()` call whereas the spec above is explicitly testing the promise that is returned from the call to `__awaiter()`. Right?

JiaLiPassion replied:
Yes, you are right, so the test code is just like the transform of the code in the comment like this

```
      zone1.run(async function (this: unknown) {
        expect(Zone.current.name).toEqual(zone1.name);
        const r = await testAsync1();
        expect(r).toEqual(4);
        expect(Zone.current.name).toEqual(zone1.name);
        expect(logs).toEqual([
          'zone1 invoke undefined',
          'before await zone1',
          'zone1 invoke native await',
          'after await1 zone1 2',
          'zone1 invoke native await',
          'after await2 zone1 4',
          'zone1 invoke native await',
        ]);
      });
```

petebacondarwin commented on packages/zone.js/test/native/native-async-await.common.spec.ts:
```diff
@@ -0,0 +1,505 @@
+  },
+});
+
+function compute(a: number) {
```
NIT: Can we rename this `increment()` and similarly `incrementAsync()` below?
It would help me to follow the logic of the tests below.

JiaLiPassion replied:
Sure, just updated.

petebacondarwin commented on packages/zone.js/test/native/native-async-await.common.spec.ts:
```diff
@@ -0,0 +1,505 @@
+         flushMicrotasks();
+         expect(log).toEqual([1, 2]);
+       }));
+  });
```
Could we have some tests that trigger the `__fakeAsyncAwaiter()` within the test block?
(Unless I am missing something).
For example:

```ts
it('should use the `__fakeAsyncAwaiter()`', fakeAsync(() => {
  let value = 0;
  const promise = Zone.__awaiter(undefined, [], function* () {
    setTimeout(() => yield value = 10, 1000);
  });
  expect(value).toBe(0);
  flushMicrotasks();
  expect(valule).toBe(10);
}));
```

JiaLiPassion replied:
Sure, I added the test cases with `Zone.__awaiter()`

petebacondarwin reviewed (commented):
Well done for sorting out the build and test configuration issues!
I have left a few comments that need to be addressed before I approve.

gkalpak commented on packages/zone.js/rollup.config.es2017.js:
```diff
@@ -0,0 +1,58 @@
+module.exports = {
+  plugins: [
+    node({
+      mainFields: ['es2017', 'module', 'jsnext:main', 'main'],
```
Do we expect there to be an `es2017` field? Should we use (or fallback to) `es2015`?

JiaLiPassion replied:
This used to bundle the test code into `es2017` to test native behavior of `await`.

gkalpak commented on packages/zone.js/lib/awaiter/zone.native.awaiter.ts:
```diff
@@ -0,0 +1,26 @@
+        res = zone.run(gen.next, gen, args, 'native await');
+        args[0] = await res.value;
+      } catch (error) {
+        zone.runGuarded(() => {
```
Why use `runGuarded()` here? What will happen if the `gen.throw()` call also throws?

JiaLiPassion replied:
Using `runGuarded()` is to trigger `zoneSpec.onHandleError` hook to let zone.js handle the error. And what do you mean `gen.throw()` also throw? `gen.throw()` will throw an error and caught by `zone.runGuarded()`.

gkalpak replied:
And what would `zoneSpec.onHandleError()` do with the error? Would it rethrow it?

JiaLiPassion replied:
`zoneSpec.onHandleError()` by default will rethrow the error, but it also supports not rethrow and swallow the error by `return false` in the `onHandleError()` hook.

gkalpak replied:
I see, thx!
I am still not 100% sure why we call `runGuarded()` here 😅
What does `zoneSpec.onHandleError()` do? When/How would it be called in the equivalent Promise-based code?

petebacondarwin replied:
A similar question... if we need `runGuarded()` here, why not on line 17 above?

petebacondarwin replied:
What would happen if we had a function that looked like:

```ts
async function foo() {
  try {
    await throw 'error 1';
  } catch (e1) {
    try {
      await throw 'error 2';
    } catch (e2) {
      doSomething();
    }
  }
}
```

Would the 2nd `try-catch` in the async function ever get executed?

gkalpak commented on packages/zone.js/lib/awaiter/zone.native.awaiter.ts:
```diff
@@ -0,0 +1,26 @@
+    const zone = Zone.current;
+    const gen = generator.apply(thisArg, _arguments);
+    let res = gen.next();
+    const args = [await res.value];
+    while (!res.done) {
+      try {
+        res = zone.run(gen.next, gen, args, 'native await');
+        args[0] = await res.value;
```
I believe there are a couple of things wrong here.
1. What if the first `res.value` (the one awaited outside the `while` loop) is a rejecting promise? We would not handle it correctly (with `gen.throw(...)`).
2. I don't think we should be awaiting `res.value` once `res.done` is true (which we currently are).

I suggest something like:

```diff
 function<T>(thisArg: any, _arguments: any, generator: () => Generator<T>) {
   const zone = Zone.current;
   const gen = generator.apply(thisArg, _arguments);
   let res = gen.next();
-  const args = [await res.value];
+  const args = [undefined];
   while (!res.done) {
     try {
+      args[0] = await res.value;
       res = zone.run(gen.next, gen, args, 'native await');
-      args[0] = await res.value;
     } catch (error) {
       ...
     }
   }
-  return args[0];
+  return res.value;
 }
```

JiaLiPassion replied:
1. I have updated the code to catch the first error.
2. About the `await` when the `res.done=true`, I think we should still `await`, since the return value maybe `Promise`,
For example,

```
 async function testAsync1() {
   let a = await increase(a);
   return Promise.resolve(a);
}
```

What do you think?

gkalpak replied:
I don't think we should await a returned promise. The promise should be returned to the caller of the function and they can do what they want with it. (AFAICT, this is also what TypeScript does.)

Note that `return Promise.resolve(a)` is different from `return await Promise.resolve(a)` and currently we are essentially treating them the same, I think.

JiaLiPassion replied:
Yeah, you are right, I updated the PR to `not await` the `res.value` when `res.done === true`.

gkalpak reviewed (changes requested):
I don't understand most of the Bazel stuff tbh, but I am not going to argue with a working setup :grin:
(Respect for figuring it out 🤘 😛)

I've left a few comments. (I haven't looked at all the tests yet.)

petebacondarwin commented on packages/zone.js/tsconfig.native.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "es2015",
```
I believe that ts_library rules will override this so it is probably redundant either way.

JiaLiPassion replied:
Yes, just like @petebacondarwin said, this will be override by `ts_library` rule, I just deleted the `target`.

JiaLiPassion commented on packages/zone.js/BUILD.bazel:
```diff
@@ -75,5 +78,7 @@ pkg_npm(
         for b in BUNDLES_ENTRY_POINTS.keys()
-    ] + [":zone_js_d_ts"],
+    ] + [":zone_js_d_ts"] + [
+        "//packages/zone.js/fesm2017:zone_awaiter",
```
Yeah, I moved this item to the array above.

JiaLiPassion commented on packages/zone.js/fesm2017/BUILD.bazel:
```diff
@@ -0,0 +1,11 @@
+    outs = [
+        "zone.awaiter.js",
+    ],
+    cmd = "tail -1 $< > temp.js && if grep -q sourceMappingURL temp.js; then head -n -1 $< > $@; else cp $< $@; fi",
```
Yes, currently the `ts_library` rule generate inline sourcemap, and it can not be removed even `inineSourceMap` is false in the `tsconfig.json`, I have confirmed with Paul, so use a shell cmd to remove it is a easier solution for now, I have added comment.

JiaLiPassion commented on packages/zone.js/lib/BUILD.bazel:
```diff
@@ -29,3 +32,25 @@ ts_library(
+    name = "awaiter",
+    srcs = [":awaiter/zone.native.awaiter.ts"],
+    devmode_target = "es2017",
+    module_name = "zone.js/lib",
```
Yes, you are right, updated.

JiaLiPassion commented on packages/zone.js/lib/BUILD.bazel:
```diff
@@ -29,3 +32,25 @@ ts_library(
+    prodmode_target = "es2017",
+    tsconfig = "//packages/zone.js:tsconfig.native.json",
+    deps = [
+        ":zone_d_ts",
+        "@npm//@types/jasmine",
+        "@npm//@types/node",
+        "@npm//rxjs",
+        "@npm//tslib",
```
Yeah, only `zone_d_ts` is needed, updated.

JiaLiPassion commented on packages/zone.js/lib/awaiter/zone.native.awaiter.ts:
```diff
@@ -0,0 +1,26 @@
+  Zone.__awaiter = async function<T>(thisArg: any, _arguments: any, generator: () => Generator<T>) {
+    const zone = Zone.current;
+    const gen = generator.apply(thisArg, _arguments);
+    let res = gen.next();
```
Yes, you are right, thank you for picking out this one, I have add `try/catch` in the `first gen.next()`.

JiaLiPassion commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -731,6 +740,37 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+      function step(result: any) {
+        result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
+      }
+      step(gen.next());
```
If `step(gen.next()` throw an error, it will automatically be resolved as a `Promise Rejection` , this is handled here https://github.com/angular/angular/blob/master/packages/zone.js/lib/common/promise.ts#L406

JiaLiPassion commented on packages/zone.js/rollup.config.es2017.js:
```diff
@@ -0,0 +1,58 @@
+// And it causes issues such as https://github.com/angular/angular/issues/40215
+// `this` should be `undefined` but is assigned with `Window` instead.
+const banner = `'use strict';
+/**
+ * @license Angular v${version}
+ * (c) 2010-2021 Google LLC. https://angular.io/
+ * License: MIT
+ */`;
```
Got it, I also updated the banner in `rollup.config.js`.

JiaLiPassion commented on packages/zone.js/rollup.config.es2017.js:
```diff
@@ -0,0 +1,58 @@
+    if (/build-esm/.test(id)) {
+      return false;
+    }
+    return /rxjs/.test(id) || /electron/.test(id);
+  },
+  output: {
+    globals: {
+      electron: 'electron',
+      'rxjs/Observable': 'Rx',
+      'rxjs/Subscriber': 'Rx',
+      'rxjs/Subscription': 'Rx',
+      'rxjs/Scheduler': 'Rx.Scheduler',
+      'rxjs/scheduler/asap': 'Rx.Scheduler',
+      'rxjs/scheduler/async': 'Rx.Scheduler',
+      'rxjs/symbol/rxSubscriber': 'Rx.Symbol',
+    },
```
Not really, this is just a `global mapping` for `rollup`, for now we only bundle the `awaiter` code to `es2017`, this maybe not necessary, but if we use this config for other code in the future, this maybe needed.

JiaLiPassion commented on packages/zone.js/BUILD.bazel:
```diff
@@ -75,5 +78,7 @@ pkg_npm(
         for b in BUNDLES_ENTRY_POINTS.keys()
-    ] + [":zone_js_d_ts"],
+    ] + [":zone_js_d_ts"] + [
+        "//packages/zone.js/fesm2017:zone_awaiter",
```
Yes, just updated!

JiaLiPassion commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -861,6 +904,14 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+  };
 }, true);
+
+Zone.__load_patch('fake_async_awaiter', (global, Zone, api) => {});
```
Yeah, I deleted this line.

JiaLiPassion commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -731,6 +740,37 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+      thisArg: any, _arguments: any, generator: () => Generator<T|Promise<T>>): Promise<T> {
+    const gen = generator.apply(thisArg, _arguments);
+    function adopt(value: any) {
+      return value && typeof value.then === 'function' ? value : new Promise(function(resolve) {
+        resolve(value);
+      });
```
I believe we can use `Promise.resolve(value)` but it will be different with the current `typescript` transformation, so basically I just keep everything the same with the `typescript` behavior.

JiaLiPassion committed:
build: update comments to remove year in the license header

Update the license header in the `rollup.config.js` of `zone.js` to

```
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
```

gkalpak commented on packages/zone.js/lib/awaiter/zone.native.awaiter.ts:
```diff
@@ -0,0 +1,35 @@
+      }
+    }
+    return args[0]!;
+  };
```
Nit: I believe this function could be simplified as suggested in https://github.com/angular/angular/pull/42768#discussion_r675610731.

JiaLiPassion replied:
@gkalpak , https://github.com/angular/angular/pull/42768#discussion_r675610731 the code you suggested here will cause `infinite loop`, when `await res.value` throw error, so the `gen.next()` will update `res` and cause `res.done` is always false.
```
 function<T>(thisArg: any, _arguments: any, generator: () => Generator<T>) {
   const zone = Zone.current;
   const gen = generator.apply(thisArg, _arguments);
   let res = gen.next(); 
    const args = [undefined];
   while (!res.done) {
     try {
      args[0] = await res.value; // If error thrown here, res will not be updated by gen.next()
       res = zone.run(gen.next, gen, args, 'native await');
     } catch (error) {
       ...
     }
   }
   return res.value;
 }
```

gkalpak replied:
Ah, good point :+1:
It is still unfortunate that we have the same bits of code before and inside the `while` loop. How about something like this (to avoid the code duplication):

```ts
function<T>(thisArg: any, _arguments: any, generator: () => Generator<T>) {
  const zone = Zone.current;
  const gen = generator.apply(thisArg, _arguments);
  let res = null;
  const args = [undefined];
  while (!res?.done) {
    try {
      res = zone.run(gen.next, gen, args, 'native await');
      args[0] = res.done ? res.value : await res.value;
    } catch (error) {
      zone.runGuarged(() => {
        res = gen.throw(error);
      }, undefined, [], 'native await');
    }
  }
  return res.value;
}
```

JiaLiPassion replied:
@gkalpak , thank you for the update, I also thought of this one, and the code above will have an additional `zone.run()` for the first `gen.next()`. So maybe this can work,

```
function<T>(thisArg: any, _arguments: any, generator: () => Generator<T>) {
  const zone = Zone.current;
  const gen = generator.apply(thisArg, _arguments);
  let res = null;
  const args = [undefined];
  while (!res?.done) {
    try {
      res = zone === Zone.current ? gen.next() : zone.run(gen.next, gen, args, 'native await');
      args[0] = res.done ? res.value : await res.value;
    } catch (error) {
      zone.runGuarged(() => {
        res = gen.throw(error);
      }, undefined, [], 'native await');
    }
  }
  return res.value;
}
```

I will update the source and test it again.

petebacondarwin commented on packages/zone.js/lib/BUILD.bazel:
```diff
@@ -29,3 +32,21 @@ ts_library(
     ],
 )
+
+ts_library(
+    name = "awaiter",
+    srcs = [":awaiter/zone.native.awaiter.ts"],
+    devmode_target = "es2017",
+    module_name = "zone.js/awaiter",
+    prodmode_target = "es2017",
+    tsconfig = "//packages/zone.js:tsconfig.native.json",
+    deps = [
+        ":zone_d_ts",
+    ],
+)
+
+filegroup(
+    name = "zone.awaiter.js",
+    srcs = [":awaiter"],
+    output_group = "es5_sources",
+)
```
If you moved this to a new bazel build file (at `packages/zone.js/lib/awaiter/BUILD.bazel`) then I think naturally you will not need the `exclude` on line 23.

JiaLiPassion replied:
Yeah, it is a great idea, but I think we still need the `exclude` on line 23 since we don't want to compile this awaiter src when build `zone.js` bundle.

petebacondarwin reviewed (changes requested):
Looking good @JiaLiPassion - thank you.
Can you add a few more tests around throwing and catching errors from within catch blocks?

gkalpak commented on packages/zone.js/lib/awaiter/zone.native.awaiter.ts:
```diff
@@ -0,0 +1,27 @@
+    const args: (T|undefined)[] = [undefined];
+    while (!res || !res.done) {
+      try {
+        res = zone === Zone.current ? gen.next() : zone.run(gen.next, gen, args, 'native await');
```
In this code, what would cause `zone` to stop being `Zone.current`?

petebacondarwin replied:
I think the idea is that `Zone.current` might change between ticks of the event loop?

JiaLiPassion replied:
In the first run, the `Zone.current` will be equal to `zone`, and after each `await` call, the `Zone.current` will become `<root>` zone.

JiaLiPassion replied:
@petebacondarwin , I am not sure why I can not reply your comment above.
```
  res = zone.run(gen.next, gen, args, 'native await');
        args[0] = await res.value;
      } catch (error) {
        zone.runGuarded(() => {
```

We can call `runGuarded` above, but we may not have the chance to call `gen.throw`, since the error will be caught by the `zone`.

JiaLiPassion replied:
@petebacondarwin , you are right, we need to `await res.value` after `gen.throw()`, I am testing the result now.

gkalpak replied:
@JiaLiPassion, is it guaranteed that `Zone.current` will be `<root>` on subsequent invocations? Can't there be any corner cases where, `Zone.current` is something else?

Similarly, is it guaranteed that `Zone.current` will not be `<root>` on the first invocation? For example, what if people run this code outside the Angular zone?

##
All in all, even if the current code works, it requires intimate Zone.js knowledge and feels brittle. I suggest we make the code more explicit by using a local variable to track whether this was the first invocation or not. (This will make the code more robust and easier to understand imo.) 

WDYT?

petebacondarwin committed:
fixup! feat(zone.js): add awaiter transform spec implementation

petebacondarwin commented:
@googlebot I consent.
